### PR TITLE
Refactored Dockerfiles

### PR DESCRIPTION
**Pull request by spara**
Changes:
- Uses alpine:3.9 as base image
- Dockerfile is a multistage build which adds glibc to alpine
- Images are unprivileged
- Added user 10151 
- configured to user 10151
- removed tini (no longer needed)
- updated Maintainer attribution

We should consider signing images as part of the build process

### CODE REVIEW

File: anaconda/Dockerfile [hunk 1]
```diff
@@ -1,32 +1,74 @@
...
-
-#  $ docker build . -t continuumio/anaconda:latest -t continuumio/anaconda:2018.12 -t continuumio/anaconda2:latest -t continuumio/anaconda2:2018.12
-#  $ docker run --rm -it continuumio/anaconda2:latest /bin/bash
-#  $ docker push continuumio/anaconda:latest
-#  $ docker push continuumio/anaconda:2018.12
-#  $ docker push continuumio/anaconda2:latest
-#  $ docker push continuumio/anaconda2:2018.12
+FROM alpine:3.9 as alpine-glibc
 
+LABEL MAINTAINER="Vlad Frolov"
```
Comment (kalefranz) [inline anaconda/Dockerfile]: What's this?

File: anaconda/Dockerfile [hunk 2]
```diff
@@ -1,32 +1,74 @@
...
+LABEL SRC=https://github.com/frol/docker-alpine-glibc
 ENV LANG=C.UTF-8 LC_ALL=C.UTF-8
-ENV PATH /opt/conda/bin:$PATH
 
-RUN apt-get update --fix-missing && apt-get install -y wget bzip2 ca-certificates \
-    libglib2.0-0 libxext6 libsm6 libxrender1 \
-    git mercurial subversion
+# Here we install GNU libc (aka glibc) and set C.UTF-8 locale as default.
+RUN ALPINE_GLIBC_BASE_URL="https://github.com/sgerrand/alpine-pkg-glibc/releases/download" && \
```
Comment (kalefranz) [inline anaconda/Dockerfile]: Are we really ok using someone else's libc?  Isn't this something we should be building ourselves?

File: anaconda3/Dockerfile [hunk 3]
```diff
@@ -1,30 +1,71 @@
...
-    echo "conda activate base" >> ~/.bashrc
+    chown -R anaconda /opt
 
-RUN apt-get install -y curl grep sed dpkg && \
-    TINI_VERSION=`curl https://github.com/krallin/tini/releases/latest | grep -o "/v.*\"" | sed 's:^..\(.*\).$:\1:'` && \
-    curl -L "https://github.com/krallin/tini/releases/download/v${TINI_VERSION}/tini_${TINI_VERSION}.deb" > tini.deb && \
-    dpkg -i tini.deb && \
-    rm tini.deb && \
-    apt-get clean
+ENV PATH "/bin:/sbin:/opt/conda/bin:/usr/bin"
```
Comment (kalefranz) [inline anaconda3/Dockerfile]: Why are we putting `/bin` and `/sbin` in front of `/opt/conda/bin`?  Also, just putting `/opt/conda/bin` on PATH isn't the proper way to activate an environment.  We should be following our own best practices.

File: anaconda3/Dockerfile [hunk 4]
```diff
@@ -1,30 +1,71 @@
...
+USER 10151
+RUN echo ". /opt/conda/etc/profile.d/conda.sh" >> ~/.bashrc && \
+    echo "conda activate base" >> ~/.bashrc
 
-ENTRYPOINT [ "/usr/bin/tini", "--" ]
-CMD [ "/bin/bash" ]
+CMD [ "bash" ]
```
Comment (kalefranz) [inline anaconda3/Dockerfile]: Where did bash come from? Alpine doesn't have bash, right?

Comment (kalefranz) [reply anaconda3/Dockerfile]: Oh, I see. We're adding it ourselves. Why? Kinda violates the spirit of alpine, no?

File: miniconda/Dockerfile [hunk 5]
```diff
@@ -1,32 +1,72 @@
...
 
-RUN wget --quiet https://repo.anaconda.com/miniconda/Miniconda2-4.5.12-Linux-x86_64.sh -O ~/miniconda.sh && \
-    /bin/bash ~/miniconda.sh -b -p /opt/conda && \
-    rm ~/miniconda.sh && \
+# Create non-root user, install dependencies, install Conda
+RUN addgroup -S anaconda && \
+    adduser -D -u 10151 anaconda -G anaconda && \
+    apk update && \
+    apk add --no-cache openssl && \
+    apk add --no-cache bash libstdc++ && \
```
Comment (kalefranz) [inline miniconda/Dockerfile]: Why do we need non-conda versions of openssl and libstdc++ hanging around? (I don't see they're being removed anywhere.) This is actually potentially dangerous if people are going to be using this as a build image base.

File: anaconda/Dockerfile (hunk 1 above)
Comment (spara) [reply anaconda/Dockerfile]: He's the maintainer of the Dockerfile that adds glibc to alpine, or are you asking about labels?

File: anaconda/Dockerfile (hunk 2 above)
Comment (spara) [reply anaconda/Dockerfile]: We could, do we have a version for this? This is also close to the example you sent me.

File: anaconda3/Dockerfile (hunk 3 above)
Comment (spara) [reply anaconda3/Dockerfile]: Sounds good, this is what the current Dockerfile has

File: anaconda3/Dockerfile (hunk 4 above)
Comment (spara) [reply anaconda3/Dockerfile]: Yes it does, adding glibc also violates the spirit of alpine

File: miniconda/Dockerfile (hunk 5 above)
Comment (spara) [reply miniconda/Dockerfile]: Kinda chicken and egg, doesn't conda need these to begin with and wget needs openssl to deal with https? If we want to use conda only packages then we can add them into bootstrap them, conda install what we need, then remove the native packages

**Comment by mingwandroid** [comment]
Good work @spara.

It would be great if we could make these images smaller (the miniconda ones that is, the anaconda ones are a lost cause size-wise).

I believe we should setup an intermediate docker image with glibc, wget, openssl, ca-cerficates and install miniconda into it, then do:

`COPY --from=intermediate /opt/conda /opt/conda`

(I got this from https://vsupalov.com/cache-docker-build-dependencies-without-volume-mounting/)

This way we can get away without installing any of glibc, wget, openssl, ca-certificates in the final image. It also means that any disk-image churn caused by installing miniconda has no effect on the final image.

I don't agree with the proposition that because we're installing glibc (which is necessary for any conda software) that makes it fine to also install something else (bash in this case). I like bash, but I feel that for the miniconda images at least we should go ultra minimal; they should contain no software that is not strictly essential. Now maybe bash is needed for too much stuff as things stand (certainly conda-build scripts assume bash), are these images to be used on things like CI? If bash is required we could have a picoconda image that is bash-less and we should seek to improve things in the ecosystem that currently require bash (post-link, activate scripts). I don't think we'll ever get to not requiring bash when it comes to building packages. For me personally, it has too many useful features that I use all the time in `build.sh`.

Anyway, I tested a little de-bashing (but not the intermediate thing!). The required changing the startup so it uses /home/anaconda/.profile. Here's my updated version of the miniconda3 Dockerfile:

```
FROM alpine:3.9 as alpine-glibc

LABEL MAINTAINER="Vlad Frolov"
LABEL SRC=https://github.com/frol/docker-alpine-glibc
ENV LANG=C.UTF-8 LC_ALL=C.UTF-8

RUN ALPINE_GLIBC_BASE_URL="https://github.com/sgerrand/alpine-pkg-glibc/releases/download" && \
    ALPINE_GLIBC_PACKAGE_VERSION="2.28-r0" && \
    ALPINE_GLIBC_BASE_PACKAGE_FILENAME="glibc-$ALPINE_GLIBC_PACKAGE_VERSION.apk" && \
    ALPINE_GLIBC_BIN_PACKAGE_FILENAME="glibc-bin-$ALPINE_GLIBC_PACKAGE_VERSION.apk" && \
    ALPINE_GLIBC_I18N_PACKAGE_FILENAME="glibc-i18n-$ALPINE_GLIBC_PACKAGE_VERSION.apk" && \
    apk add --no-cache --virtual=.build-dependencies wget ca-certificates && \
    echo \
        "-----BEGIN PUBLIC KEY-----\
        MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEApZ2u1KJKUu/fW4A25y9m\
        y70AGEa/J3Wi5ibNVGNn1gT1r0VfgeWd0pUybS4UmcHdiNzxJPgoWQhV2SSW1JYu\
        tOqKZF5QSN6X937PTUpNBjUvLtTQ1ve1fp39uf/lEXPpFpOPL88LKnDBgbh7wkCp\
        m2KzLVGChf83MS0ShL6G9EQIAUxLm99VpgRjwqTQ/KfzGtpke1wqws4au0Ab4qPY\
        KXvMLSPLUp7cfulWvhmZSegr5AdhNw5KNizPqCJT8ZrGvgHypXyiFvvAH5YRtSsc\
        Zvo9GI2e2MaZyo9/lvb+LbLEJZKEQckqRj4P26gmASrZEPStwc+yqy1ShHLA0j6m\
        1QIDAQAB\
        -----END PUBLIC KEY-----" | sed 's/   */\n/g' > "/etc/apk/keys/sgerrand.rsa.pub" && \
    wget \
        "$ALPINE_GLIBC_BASE_URL/$ALPINE_GLIBC_PACKAGE_VERSION/$ALPINE_GLIBC_BASE_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_BASE_URL/$ALPINE_GLIBC_PACKAGE_VERSION/$ALPINE_GLIBC_BIN_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_BASE_URL/$ALPINE_GLIBC_PACKAGE_VERSION/$ALPINE_GLIBC_I18N_PACKAGE_FILENAME" && \
    apk add --no-cache \
        "$ALPINE_GLIBC_BASE_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_BIN_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_I18N_PACKAGE_FILENAME" && \
    \
    rm "/etc/apk/keys/sgerrand.rsa.pub" && \
    /usr/glibc-compat/bin/localedef --force --inputfile POSIX --charmap UTF-8 "$LANG" || true && \
    echo "export LANG=$LANG" > /etc/profile.d/locale.sh && \
    \
    apk del glibc-i18n && \
    \
    rm "/root/.wget-hsts" && \
    apk del .build-dependencies && \
    rm \
        "$ALPINE_GLIBC_BASE_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_BIN_PACKAGE_FILENAME" \
        "$ALPINE_GLIBC_I18N_PACKAGE_FILENAME"

FROM alpine-glibc

LABEL MAINTAINER="Anaconda, Inc"
ENV CONDA_VERSION 4.5.12

# Create non-root user, install dependencies, install Conda
RUN addgroup -S anaconda && \
    adduser -D -u 10151 anaconda -G anaconda && \
    wget --quiet https://repo.continuum.io/miniconda/Miniconda3-$CONDA_VERSION-Linux-x86_64.sh && \
    mv Miniconda3-$CONDA_VERSION-Linux-x86_64.sh miniconda.sh && \
    mkdir -p /opt && \
    sh ./miniconda.sh -b -p /opt/conda && \
    rm miniconda.sh && \
    ln -s /opt/conda/etc/profile.d/conda.sh /etc/profile.d/conda.sh && \
    chown -R anaconda:anaconda /opt && \
    echo ". /opt/conda/etc/profile.d/conda.sh" >> /home/anaconda/.profile && \
    echo "conda activate base" >> /home/anaconda/.profile

USER  10151
ENV PATH "/bin:/sbin:/usr/bin"

CMD [ "sh", "--login", "-i" ]
```

**Comment by spara** [comment]
Changes look good, the only thing is that a user will need to source .profile to run conda, is this the intended behavior?

**Comment by soapy1** [comment]
This looks good to me! Could you also update some of the README's also. For example https://github.com/ContinuumIO/docker-images/blob/master/anaconda/README.md#usage doesn't work with the jupyter notebook example anymore.

**Comment by mingwandroid** [comment]
Taking a quick look at this now, I will try to add the `COPY --from=intermediate` stuff. I may also investigate UPX packing every binary and shared library .. maybe for a picoconda?

**Pull request merged**